io-sim/test/Test/Control/Monad/Common.hs

```diff
@@ -0,0 +1,509 @@
+               | a  < 0 && b  < 0 = EQ
+               | otherwise = a `compare` b
+
+--
+-- Forking
+--
+
+test_fork_order :: forall m.
+                   ( MonadFork m
+                   , MonadSTM m
+                   , MonadTimer m
+                   )
+                => Positive Int
+                -> m Property
+test_fork_order = \(Positive n) -> isValid n <$> withProbe (experiment n)
+  where
+    experiment :: Int -> Probe m Int -> m ()
+    experiment 0 _ = return ()
+    experiment n p = do
+      v <- newTVarIO False
+
+      void $ forkIO $ do
+        probeOutput p n
+        atomically $ writeTVar v True
+      experiment (n - 1) p
+
+      -- wait for the spawned thread to finish
+      atomically $ readTVar v >>= check
+
+    isValid :: Int -> [Int] -> Property
+    isValid n tr = tr === [n,n-1..1]
+
+test_threadId_order :: forall m.
+                   ( MonadFork m
+                   , MonadSTM m
+                   , MonadTimer m
+                   )
+                => Positive Int
+                -> m Property
+test_threadId_order = \(Positive n) -> do
+    isValid n <$> (forM [1..n] (const experiment))
+  where
+    experiment :: m (ThreadId m)
+    experiment = do
+      v <- newTVarIO False
+
+      tid <- forkIO $ atomically $ writeTVar v True
+
+      -- wait for the spawned thread to finish
+      atomically $ readTVar v >>= check
+      return tid
+
+    isValid :: Int -> [ThreadId m] -> Property
+    isValid n tr = map show tr === map (("ThreadId " ++ ) . show . (:[])) [1..n]
+
+-- This property is not actually deterministic in IO. Uncomment the following
+-- and try it! Arguably therefore, this property does not need to be true for
+-- the Sim either. Perhaps we should introduce random scheduling and abandon
+-- this property. In the meantime it's a helpful sanity check.
+
+--prop_wakeup_order_IO = ioProperty test_wakeup_order
+
+test_wakeup_order :: ( MonadFork m
+                     , MonadSTM m
+                     , MonadTimer m
+                     )
+                => m Property
+test_wakeup_order = do
+    v          <- newTVarIO False
+    wakupOrder <-
+      withProbe $ \p -> do
+        sequence_
+          [ do _ <- forkIO $ do
+                 atomically $ do
+                   x <- readTVar v
+                   check x
+                 probeOutput p (n :: Int)
+               threadDelay 0.1
+          | n <- [0..9] ]
+        atomically $ writeTVar v True
+        threadDelay 0.1
+    return (wakupOrder === [0..9]) --FIFO order
+
+--
+-- Probe mini-abstraction
+--
+
+-- | Where returning results directly is not convenient, we can build up
+-- a trace of events we want to observe, and can do probe output from
+-- multiple threads.
+--
+type Probe m x = StrictTVar m [x]
+
+withProbe :: MonadSTM m => (Probe m x -> m ()) -> m [x]
+withProbe action = do
+    probe <- newTVarIO []
+    action probe
+    reverse <$> atomically (readTVar probe)
```

Comment:
we used this pattern when we didn't had `Dynamic` tracing ...

Reply:
I didn't introduce this, I just refactored the code. If this needs to change maybe it should be done in a different PR

Reply:
I know, I am just explaining why it's there.